### ISSUE TITLE
7.12 Extensions dictionary's BaseVersion limits extension applicability

### ISSUE THREAD

**Issue by bsanchezb** [pdf-association/pdf-issues#399]
**Describe the bug**

There are certain standards defining new dictionaries or extending PDF's functionality, such as ISO 32001-2022, ISO 32002-2022, etc. and last but not least [ETSI EN 319 142-1](https://www.etsi.org/deliver/etsi_en/319100_319199/31914201/01.02.01_60/en_31914201v010201p.pdf) for PAdES signatures.

The latter defines a few extension dictionaries to be used for incorporation of /DSS dictionary or a use of ETSI.CAdES.detached SubFilter, as follows:
```
<</ESIC 
       <</BaseVersion /1.7  
          /ExtensionLevel 1 
        >> 
    >>
```
and
```
<</ESIC 
       <</BaseVersion /1.7  
          /ExtensionLevel 2 
        >> 
    >> 
```
respectively.

Both extensions contain a /BaseVersion field with a value "1.7".

ISO 32000-1 (and ISO 32000-2 as well) has the following description of the attribute:

> 7.12.3 BaseVersion 
The value of the BaseVersion entry shall be a name and shall be consistent with the syntax used for 
the Version entry value of the catalog dictionary (see 7.7.2, "Document catalog dictionary"). **The value 
of BaseVersion**, when treated as a version number, **shall be less than or equal to the PDF version**, **both 
in the document header** (see 7.5.2, "File header") and **the catalog dictionary Version** key value, if 
present. The value of BaseVersion may be different from the version number in the document header 
or that supplied by the Version key in the catalog dictionary. This is because it reflects the version of 
the standard that has been extended and not the version of this particular file. 

Which basically means that one is allowed to use the aforementioned extensions **only for the PDF document of version 1.7 or higher**, as overwriting of /Version value within a **catalog dictionary** does not bring any benefit (as both values shall match the requirement, so overwriting does not make sense as document header does not change).

This makes the extensions with a BaseVersion 1.7 not applicable in practice for PDF documents produced with a lower version (e.g. 1.4, 1.5, etc.).

The same applies for ISO 32001 and ISO 32002, defining extension dictionary entries with a /BaseVersion 2.0, making them inapplicable for older PDF documents.

This brings a question, whether inclusion of the extension thus breaking the /BaseVersion's requirement **or** omitting the extension while applying certain customizations from the referenced standards is actually **better for interoperability** of the PDF document produced with a lower version?

Not applying the standards at all, such as ETSI EN 319 142-1, is practically impossible for creation of valid PAdES LTV signatures.

Therefore, what would be the best way to proceed with the aforementioned /Extensions dictionaries? Maybe the ISO 32000-1/-2 does not define what it intends to?

**Additional context**
The initial question about the /Extensions dictionary was brought at EU eSig DSS JIRA (see [DSS-2392](https://ec.europa.eu/digital-building-blocks/tracker/browse/DSS-2392)).

**Comment by petervwyatt** [comment]
The DigSig TWG will discuss and propose a solution - @MatthiasValvekens

**Comment by mkl-public** [comment]
While the extension mechanism restrictions are especially relevant in the signing context (as one cannot simply change the original PDF version of signed PDFs), the underlying question is of broader interest. So IMO discussion should not be limited to the DigSig TWG: Being limited to incremental updates is most relevant in the context of signing but can also occur in other contexts. (Unfortunately I could not attend last night's PDF TWG meeting and, therefore, could not make that point there.)

Essentially the current formulation is good for the use case

> _I create a document from scratch (or have an existing document I can arbitrarily change) and want to add content defined in an extension._

But it's not good for the use case 

> _I have a document I can only change by incremental updates, and want to add content defined in an extension._

If the version number in my source document is lower than my extension base version, the spec currently does not allow applying the extension, even if I put a high enough **Version** into the **Catalog**.

Thus, is there any reason not to change the sentence quoted by @bsanchezb 

>The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the PDF version, both in the document header (see 7.5.2, "File header") and the catalog dictionary **Version** key value, 

to something like

>The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the PDF version, either in the document header (see 7.5.2, "File header") or the catalog dictionary **Version** key value, if present,

**Comment by bsanchezb** [comment]
Just to add more context on this, the chapter "7.5.6 Incremental updates" defines:

> In versions of PDF 1.4 or later a PDF writer **may use** the **Version** entry in the document’s catalog 
dictionary (see 7.7.2, "Document catalog dictionary") **to override the version specified in the header**. A 
PDF writer may also need to update the Extensions dictionary, see 7.12, "Extensions dictionary", if the 
update either deleted or added developer-defined extensions.

Which basically implies that /Version field within the catalog is meant to override the version in the document header. Thus, having the requirement of "7.12.3 BaseVersion" for **both** document header and catalog /Version key makes little sense to me.

I would go even further, than the proposal of @mkl-public, and simplify it to:

>The value of BaseVersion, when treated as a version number, shall be less than or equal to the PDF version.

Which would imply the document header version, or catalog's /Version key value, when present. Especially, considering the fact that the value key of catalog's /Version field cannot be lower than the version defined in the document header, based on the definition.

But of course, I will leave it to the experts :)

**Comment by petervwyatt** [comment]
Let's start in DigSig TWG as there are likely other issues needing fixing that will also be in play here. 
PDF TWG will then review any and all recommendations.

I also agree re the "term of art" suggestion above. And noting that the DocCatalog **Version** entry cannot downgrade a PDF (I don't think that is explicitly stated anywhere). However anything that relies on version number is going to be fraught with issues!! See [my recent article](https://pdfa.org/pdf-versions/) (shameless self-promotion). Furthermore, the Extensions dictionary was ONLY introduced with ISO 32000-1:2008 - it was never defined in any Adobe specification (including Adobe PDF 1.7).

**Comment by bsanchezb** [comment]
>  And noting that the DocCatalog Version entry cannot downgrade a PDF (I don't think that is explicitly stated anywhere).

In fact it is defined in the standard:

> **Table 29 — Entries in the catalog dictionary**
**Version** (Optional; PDF 1.4) The version of the PDF specification to which 
the document conforms (for example, 1.4) **if later than the version 
specified in the file’s header** (see 7.5.2, "File header"). If the header 
specifies a later version, or if this entry is absent, the document 
shall conform to the version specified in the header. This entry 
enables a PDF processor to update the version using an incremental 
update; see 7.5.6, "Incremental updates". 

So basically yes, only the later (or higher) version, than the one in file's header, is allowed in catalog's /Version field.

@MatthiasValvekens , thank you for your reply. If I understand well, the current definition of a /BaseVersion for the developer extensions dictionary is not correct per se. Thus, one may ignore the requirement and **include the needed extensions**, provided that **the current version of the document allows so** (i.e. either version defined in the file's header, or in the catalog's /Version field, when present). Will it be ok to follow this logic, or do we need to wait for an update in the standard?

**Comment by mkl-public** [comment]
> the current definition of a /BaseVersion for the developer extensions dictionary is not correct per se.

Well, I'd say it's not like we want it, in particular in the context of signed files, but it's not a question of correctness.

What bothers me a bit is that someone at the time of creating that definition explicitly formulated it referencing both the header version and the Catalog version. This seems to indicate that person saw a reason for that. If that person is reading here or is known to someone reading here, it would be interesting to get to know the reasoning.

**Comment by petervwyatt** [comment]
Re my downgrade comment: sorry, but I wasn't clear enough...

The current wording only references the header and not the "term of art" as @MatthiasValvekens suggests: "_if later than the version specified in the file’s header_". This does not account for multiple incremental updates which can each update the DocCatalog and Extensions dictionaries. For example, if we have a PDF with a header of `%PDF-1.4` and with a 1st (older) incremental update of `/Version /1.7` can a 2nd (later) incremental update change that back to `/Version /1.5`? That seems wrong to me - or is there a valid use-case? Or doesn't it matter? The 1st incremental update may be adding a digital signature and Extension dict, and then the 2nd incremental update updates just the Extension Dictionary along with other stuff (so the dig-sig is validatable only against the PDF revision with just the 1st incremental update). Obviously the header cannot be changed as that would invalidate the digsig, and likely why /Version exists in the first place. Hence why I wanted the DigSig TWG to research and discuss all the issues and come back with what I assume will be a long list of spec improvements.

**Comment by mkl-public** [comment]
> For example, if we have a PDF with a header of `%PDF-1.4` and with a 1st (older) incremental update of `/Version /1.7` can a 2nd (later) incremental update change that back to `/Version /1.5`?

Indeed, you are right, this case is not specifically covered in the spec.

> That seems wrong to me - or is there a valid use-case?

I also think that you should only be allowed to increase (or keep) the version, never to decrease it. 

And neither do I see a valid use case, merely lazy implementations ("I need version 1.6, so let's put `/Version /1.6` into the **Catalog**, no matter what the effective version of the document was before.").

> Or doesn't it matter?

Admittedly, while in theory the **Extensions** mechanism (with the issue discussed here clarified) sounds nice, there is too much software out there that doesn't add **Version** or **DeveloperExtensions** entries at all or that overwrites existing entries with their own values, no matter what was there before. Furthermore, in ETSI PAdES specs adding extensions only is recommended, not required, making not adding extensions an easy way out.

Are there any _required_ **Extensions** entries anywhere? I'm only aware of those in the ISO TS documents 32001 etc..

> Hence why I wanted the DigSig TWG to research and discuss all the issues and come back with what I assume will be a long list of spec improvements.

Well, that sounds like a long way to go.

This is no problem for the PAdES people as the extension there currently is optional.

But what about implementors of ISO TS 32001 etc.? If you want those extensions to be in use before the EOL of the TSs, a way to use them in incrementally updated documents with PDF-1.x headers should be presented soon.

**Comment by MatthiasValvekens** [comment]
DigSig TWG sees this as an issue with the extensions dictionary, not a DigSig problem -> WG8.

**Comment by mkl-public** [comment]
Reconsidering what has been said on the DigSig meeting some hours ago I have a question: If I recall correctly, @lrosenthol said that the extension mechanism as is has explicitly been designed with the restriction that's the topic here. Here I wonder, _why_?

In detail:

The *Version* entry is specified as _The version of the PDF specification to which the document conforms (for example, 1.4) if later than the version specified in the file’s header (see 7.5.2, "File header")._

So if I have a document with a 1.4 header version and a 1.7 catalog version (in an incremental update), the document must conform to 1.7. Why then is the extension mechanism explicitly designed to not allow the document to have extensions with a 1.7 base version?

**Comment by petervwyatt** [comment]
The quick answer is that all ISO/TS extensions and clarification publications to date only apply to ISO 32000-2 (PDF 2.0) and are **NOT** retrospective to ISO 32000-1:2008 (PDF 1.7). So the Extensions dictionary that was introduced with ISO 32000-1:2008 (PDF 1.7) can have a **BaseVersion** of _1.7_ for some things that work with PDF 1.7 (like the Adobe Extensions Level 3 and 5), BUT all the ISO/TS specifications are intentionally limiting themselves to PDF 2.0 by those specifications requiring **BaseVersion** to be 2.0.

PS. This means it is NOT valid to simply put STEP, glTF, AES-GCM 256, SHA-3 and friends, etc. into PDF 1.7. You need to first "upgrade" the PDF by adding a new DocCatalog **Version** 2.0 entry - but that is all that is needed to do an "upgrade".

**Comment by MatthiasValvekens** [comment]
> You need to first "upgrade" the PDF by adding a new DocCatalog Version 2.0 entry - but that is all that is needed to do an "upgrade".

Yeah, I think that's the intended behaviour, but the letter of the spec (on BaseVersion) currently doesn't allow this, so we'd need to fix the requirement that BaseVersion also be less than or equal to the header version for this to work.

**Comment by petervwyatt** [comment]
Trying to converge to a solution...

The problematic sentence is "_The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the PDF version, both in the document header (see 7.5.2, "File header") and the catalog dictionary **Version** key value, if present._" with interpretation of the word "_both_". 

So borrowing phrasing from Table 29:
"_The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the version of the PDF specification to which this document conforms (see "Table 29 - Entries in the catalog dictionary", considering both the document header (see 7.5.2, "File header") and the catalog dictionary **Version** key value", if present)_"

**Comment by mkl-public** [comment]
Sounds good. 

In particular if we somewhere resolve the issue you brought up, so that later catalog version entry cannot decrease a former catalog version entry anymore.

**Comment by petervwyatt** [comment]
@mkl-public thanks for the reminder...

I think the easiest place to do this is the paragraph above Note 4 in 7.5.6 Incremental updates:

> In versions of PDF 1.4 or later a PDF writer may use the **Version** entry in the document’s catalog dictionary (see 7.7.2, "Document catalog dictionary") to ~~override the version specified in the header~~ **upgrade the current version of the PDF specification to which the document conformed (considering both the document header (see 7.5.2, "File header") and the catalog dictionary Version key value, if already present). When updating a trailer, a PDF writer shall not reduce the version of a document.** A PDF writer may also need to update the Extensions dictionary, see 7.12, "Extensions dictionary", if the update either deleted or added developer-defined extensions.
>NOTE 4 The **Version** entry enables the version to be ~~altered~~ **upgraded** when performing an incremental update.

**Comment by mkl-public** [comment]
>When updating a trailer, a PDF writer shall not reduce the version of a document.

Maybe this should be formulated as a file format requirement, not a file processor requirement, along the lines of "a catalog of an incremental update must not reduce the version of its document with the value or absence of its Version entry."

**Comment by mkl-public** [comment]
As in the TWG the exact proposed solution was unclear, it consists of two parts:

1. The first part addresses the original question, and the proposed change is:

   > The problematic sentence is "_The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the PDF version, both in the document header (see 7.5.2, "File header") and the catalog dictionary **Version** key value, if present._" with interpretation of the word "_both_".
   > 
   > So borrowing phrasing from Table 29: "_The value of **BaseVersion**, when treated as a version number, shall be less than or equal to the version of the PDF specification to which this document conforms (see "Table 29 - Entries in the catalog dictionary", considering both the document header (see 7.5.2, "File header") and the catalog dictionary **Version** key value", if present)_"

2. The second part addresses the issue pointed out by Peter that catalog **Version** entries in later incremental updates are not forbidden yet to downgrade catalog **Version** from earlier incremental updates. The proposed change is:

   > I think the easiest place to do this is the paragraph above Note 4 in 7.5.6 Incremental updates:
   > 
   >> In versions of PDF 1.4 or later a PDF writer may use the **Version** entry in the document’s catalog dictionary (see 7.7.2, "Document catalog dictionary") to ~override the version specified in the header~ **upgrade the current version of the PDF specification to which the document conformed (considering both the document header (see 7.5.2, "File header") and the catalog dictionary Version key value, if already present). A catalog of an incremental update must not reduce the version of its document with the value or absence of its Version entry.** A PDF writer may also need to update the Extensions dictionary, see 7.12, "Extensions dictionary", if the update either deleted or added developer-defined extensions.
   >> NOTE 4 The **Version** entry enables the version to be ~altered~ **upgraded** when performing an incremental update.

**Comment by petervwyatt** [comment]
Will review the final wording of both the changes that @mkl-public kindly summarized in the previous comment in the next PDF TWG meeting. Leaving in "proposed solution" state.
